robosat/osm/road.py

```diff
@@ -0,0 +1,179 @@
+            "lane_width": 3.50,
+            "left_hard_shoulder_width": 0.00,
+            "right_hard_shoulder_width": 0.75,
+        },
+        "unclassified": {
+            "lanes": 1,
+            "lane_width": 3.50,
+            "left_hard_shoulder_width": 0.00,
+            "right_hard_shoulder_width": 0.00,
+        },
+        "residential": {
+            "lanes": 1,
+            "lane_width": 3.50,
+            "left_hard_shoulder_width": 0.00,
+            "right_hard_shoulder_width": 0.75,
+        },
+        "service": {
+            "lanes": 1,
+            "lane_width": 3.00,
+            "left_hard_shoulder_width": 0.00,
+            "right_hard_shoulder_width": 0.00,
+        },
+        "motorway_link": {
+            "lanes": 2,
+            "lane_width": 3.75,
+            "left_hard_shoulder_width": 0.75,
+            "right_hard_shoulder_width": 3.00,
+        },
+        "trunk_link": {
+            "lanes": 2,
+            "lane_width": 3.75,
+            "left_hard_shoulder_width": 0.50,
+            "right_hard_shoulder_width": 1.50,
+        },
+        "primary_link": {
+            "lanes": 1,
+            "lane_width": 3.50,
+            "left_hard_shoulder_width": 0.00,
+            "right_hard_shoulder_width": 0.75,
+        },
+        "secondary_link": {
+            "lanes": 1,
+            "lane_width": 3.50,
+            "left_hard_shoulder_width": 0.00,
+            "right_hard_shoulder_width": 0.75,
+        },
+        "tertiary_link": {
+            "lanes": 1,
+            "lane_width": 3.50,
+            "left_hard_shoulder_width": 0.00,
+            "right_hard_shoulder_width": 0.00,
+        },
+    }
+
+    road_filter = set(highway_attributes.keys())
+
+    EARTH_MEAN_RADIUS = 6371004.0
+
+    def __init__(self):
+        super().__init__()
+        self.features = []
+
+    def way(self, w):
+        if "highway" not in w.tags:
+            return
+
+        if w.tags["highway"] not in self.road_filter:
+            return
+
+        left_hard_shoulder_width = self.highway_attributes[w.tags["highway"]][
+            "left_hard_shoulder_width"
+        ]
+        lane_width = self.highway_attributes[w.tags["highway"]]["lane_width"]
+        lanes = self.highway_attributes[w.tags["highway"]]["lanes"]
+        right_hard_shoulder_width = self.highway_attributes[w.tags["highway"]][
+            "right_hard_shoulder_width"
+        ]
+
+        if "oneway" not in w.tags:
+            lanes = lanes * 2
+        elif w.tags["oneway"] == "no":
+            lanes = lanes * 2
+
+        if "lanes" in w.tags:
+            try:
+                lanes = int(w.tags["lanes"])
+            except ValueError as e:
+                print(
+                    "ValueError: ",
+                    e,
+                    " Invalid road lanes property: https://www.openstreetmap.org/way/{}".format(
+                        w.id
+                    ),
+                    file=sys.stderr,
+                )
```

Comment:
Maybe don't print ValueError and the exception e here; that's not really useful for users. What about:


```
print("Warning: invalid lane tag: https://www.openstreetmap.org/way/{}".format(w.id), file=sys.stderr)
```

Reply:
I will pay my attention on post-process for roads later on. Would love to talk about it.

Reply:
Sure, feel free to open a ticket for it. A rough idea could be to dijkstra through the segmentation probabilities; maybe trying the opencv skeletonization; doing some simplification. Interesting problem for sure!